lib/commands/push.ts

```diff
@@ -234,146 +258,165 @@ export default class PushCmd extends Command {
 		const {
-			validateApplicationName,
 			validateDotLocalUrl,
 			validateIPAddress,
 		} = await import('../utils/validation');
 
-		// First try the application regex from the api
-		if (validateApplicationName(appOrDevice)) {
-			return BuildTarget.Cloud;
-		}
 
 		if (validateIPAddress(appOrDevice) || validateDotLocalUrl(appOrDevice)) {
 			return BuildTarget.Device;
 		}
 
-		return null;
+		return BuildTarget.Cloud;
 	}
 
-	async getAppOwner(sdk: BalenaSDK, appName: string) {
-		const _ = await import('lodash');
+	async remoteBuild(
+		application: string,
+		dockerfilePath: string,
+		registrySecrets: RegistrySecrets,
+		options: FlagsDef,
+		sdk: BalenaSDK,
+	) {
+		const remote = await import('../utils/remote-build');
+
+		// Check for invalid options
+		const localOnlyOptions = ['nolive', 'service', 'system', 'env'];
+
+		localOnlyOptions.forEach((opt) => {
+			// @ts-ignore : Not sure why typescript wont let me do this?
+			if (options[opt]) {
+				throw new ExpectedError(
+					`The --${opt} flag is only valid when pushing to a local mode device`,
+				);
+			}
+		});
+
+		await Command.checkLoggedIn();
+		const [token, baseUrl, owner] = await Promise.all([
+			sdk.auth.getToken(),
+			sdk.settings.get('balenaUrl'),
+			this.getAppOwnerOrg(sdk, application),
+		]);
+
+		const opts = {
+			dockerfilePath,
+			emulated: options.emulated,
+			multiDockerignore: options['multi-dockerignore'],
+			nocache: options.nocache,
+			registrySecrets,
+			headless: options.detached,
+			convertEol: !options['noconvert-eol'],
+		};
+		const args = {
+			app: application,
+			owner,
+			source: options.source,
+			auth: token,
+			baseUrl,
+			nogitignore: !options.gitignore,
+			sdk,
+			opts,
+		};
+		await remote.startRemoteBuild(args);
+	}
+
+	async localDeploy(
+		localDeviceAddress: string,
+		dockerfilePath: string,
+		registrySecrets: RegistrySecrets,
+		options: FlagsDef,
+	) {
+		const deviceDeploy = await import('../utils/device/deploy');
+
+		// TODO: Support passing a different port
+		try {
+			await deviceDeploy.deployToDevice({
+				source: options.source,
+				deviceHost: localDeviceAddress,
+				dockerfilePath,
+				registrySecrets,
+				multiDockerignore: options['multi-dockerignore'],
+				nocache: options.nocache,
+				pull: options.pull,
+				nogitignore: !options.gitignore,
+				noParentCheck: options['noparent-check'],
+				nolive: options.nolive,
+				detached: options.detached,
+				services: options.service,
+				system: options.system,
+				env: options.env || [],
+				convertEol: !options['noconvert-eol'],
+			});
+		} catch (e) {
+			const { BuildError } = await import('../utils/device/errors');
+			if (instanceOf(e, BuildError)) {
+				throw new ExpectedError(e.toString());
+			} else {
+				throw e;
+			}
+		}
+	}
 
+	async getAppOwnerOrg(sdk: BalenaSDK, appNameOrSlug: string) {
+		// If it's a slug, extract org and return
+		if(appNameOrSlug.includes('/')) {
+			return appNameOrSlug.split('/')[0];
```

Comment:
Hmm, this isn't really validation though, more like just detection that they meant to specify a slug.   I don't think proper validation would be useful here because (if I understand correctly) slugs can in some cases (openBalena) not include `/`.   I think a 'looksLikeSlug()' function would be good though.

Reply:
> I think a 'looksLikeAppSlug()' function would be good though.

Actually I think in this context that wouldn't be the right way to go, since the next line (408) also has knowledge about the `/`.

Reply:
It's just that I expect that, every now and then, users (especially beginners) will use the org name instead of the org handle when providing the slug, e.g. `balena push MyOrg/MyApp` or `balena push Fundação/app3`, and a basic regex check for characters outside the handle range `[a-z_]` (needs confirmation) could provide a friendlier message such as:
```
Invalid organization handle. Was the organization name used instead of its handle?
```
The default error message appears to be:
```
$ ./bin/balena push MyOrg/MyApp -s ~/balena/other/simple_container/
[Error]    Failed to fetch information about given application.
[Error]    Not deploying release.
Remote build failed
```
By the way, it seems that it does not work even when I use the correct handle as reported by `balena apps`...
```
$ ./bin/balena push gh_paulo_castro/test-nuc -s ~/balena/other/simple_container/
[Error]    Failed to fetch information about given application.
[Error]    Not deploying release.
Remote build failed

$ ./bin/balena whoami
== ACCOUNT INFORMATION
USERNAME: gh_paulo_castro
EMAIL:    pefcastro@gmail.com
URL:      balena-cloud.com

$ ./bin/balena orgs
NAME                           HANDLE
gh_paulo_castro's Organization gh_paulo_castro
...

$ ./bin/balena apps
ID      APP NAME         SLUG                            DEVICE TYPE     ONLINE DEVICES DEVICE COUNT
1491721 test-nuc         gh_paulo_castro/test-nuc        intel-nuc       0              1
...

$ ./bin/balena push test-nuc -s ~/balena/other/simple_container/
[Info]     Starting build for test-nuc, user gh_paulo_castro
...
[Success]  Release successfully created!
...
[Info]     Build finished in 23 seconds
```

Reply:
> It's just that I expect that, every now and then, users (especially beginners) will use the org name instead of the org handle when providing the slug, e.g. balena push MyOrg/MyApp or balena push Fundação/app3

Other commands that accept slugs use `lowercaseIfSlug` during the parsing, which allows these to work most of the time (this was done on Akis' recommendation).    That wouldn't work with Fundação/app3, but that will throw an error as expected.  We should probably do the same here.

> The default error message appears to be: `Failed to fetch information about given application`

Right we should improve that.

> By the way, it seems that it does not work even when I use the correct handle as reported by balena apps...

That's very strange, since I had tested.  Will investigate.